### PR TITLE
#166265908 add new books

### PR DESCRIPTION
**Pull request by PelumiAlesh**
### What does this PR do?
Create `POST`  request on add book endpoint
#### Description of Task to be completed?

- Create endpoints for add book

- User should be able to add book

- Test for add books endpoint should be written

- Validations for add book should be added

#### How should this be manually tested?
After cloning the repo

1. Run `npm run migrate`

2. Then run `npm run dev`

3. Now make a post request to `localhost:3000/api/v1/books`

#### Any background context you want to provide?
The authors table is yet to be created
#### What are the relevant pivotal tracker stories?
[#166265907](https://www.pivotaltracker.com/story/show/166265908)
#### Screenshots (if appropriate)
Screenshot 2019-06-05 at 9 13 29 PM

#### Questions:
No

### CODE REVIEW

**Comment by chuxmykel** [comment]
#### Now make a post request to `localhost:3000/api/v1/book`
I think `/books` would be more appropriate

File: server/__test__/book.spec.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import debug from 'debug';
+import supertest from 'supertest';
+import app from '../index';
+import mockBook from './mock_data/mock_books';
+
+const log = debug('dev');
```
[the remark this answers is not in the document]
Comment (chuxmykel) [reply server/__test__/book.spec.js]: @PelumiAlesh Was this actually resolved?

File: server/db/migrations/20190605110045-create-book.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      allowNull: false,
+      type: Sequelize.DATE
+    },
+    updatedAt: {
+      allowNull: false,
+      type: Sequelize.DATE
+    }
+  });
+}
+export function down(queryInterface, Sequelize) {
```
[the remark this answers is not in the document]
Comment (chuxmykel) [reply server/db/migrations/20190605110045-create-book.js]: Remove the Sequelize parameter from the down function. It's unnecessary there.

**Review by encodedBicoding — COMMENTED** [review]
I second Chux thoughts

File: server/docs/hellobooks_api_doc.yaml [hunk 3]
```diff
@@ -17,12 +17,13 @@ externalDocs:
 
 servers:
 - url: /api/v1
-- url: https://hello-books-lib-staging.herokuapp.com/api/v1/ # Heroku deployment (staging)
+- url: localhost:6000/api/v1/ # Heroku deployment (staging)
```
Comment (chuxmykel) [inline server/docs/hellobooks_api_doc.yaml]: Why did you change this?

File: server/routes/index.js [hunk 4]
```diff
@@ -15,5 +16,6 @@ router.get('/', (req, res) => {
 
 router.use('/auth', userRouter);
 router.use('/oauth', socialRouter);
+router.use('/book', bookRouter);
```
Comment (chuxmykel) [inline server/routes/index.js]: I think `books` would be more appropriate.

**Review by chuxmykel — CHANGES_REQUESTED** [review]

File: server/routes/index.js (hunk 4 above)
Comment (OvieMudi) [reply server/routes/index.js]: I agree.

File: server/db/migrations/20190605110045-create-book.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (chuxmykel) [reply server/db/migrations/20190605110045-create-book.js]: @PelumiAlesh

File: server/controllers/bookController.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      const book = await models.Books.create({
+        title,
+        description,
+        amount,
+        authorID,
+        status,
+        year,
+      });
+
+      return Utils.successStatus(res, 201, 'Book added successfully', {
```
Comment (chuxmykel) [inline server/controllers/bookController.js]: Return the book object directly here. There's no need of calling out each property since you're sending all of them back in the response

**Review by chuxmykel — DISMISSED** [review]

File: .eslintrc [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "root": true,
-  "extends": ["airbnb-base", "prettier"],
+"extends": ["airbnb-base", "prettier"],
```
Comment (emp-daisy) [inline .eslintrc]: What's happening here?

File: server/db/migrations/20190605110045-create-book.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    description: {
+      allowNull: false,
+      type: Sequelize.TEXT
+    },
+    amount: {
+      allowNull: false,
+      type: Sequelize.NUMERIC
+    },
+    authorID: {
+      allowNull: false,
```
Comment (emp-daisy) [inline server/db/migrations/20190605110045-create-book.js]: What happens if the author is unknown?

File: server/controllers/bookController.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+      authorID,
+      status,
+      year
+    } = req.body;
+    try {
+      const book = await models.Books.create({
+        title,
+        description,
+        amount,
+        authorID,
```
Comment (emp-daisy) [inline server/controllers/bookController.js:21]: Are you validating the author's ID is correct?

File: .eslintrc (hunk 6 above)
Comment (PelumiAlesh) [reply .eslintrc]: Nothing, its the same as before

File: server/db/migrations/20190605110045-create-book.js (hunk 7 above)
Comment (PelumiAlesh) [reply server/db/migrations/20190605110045-create-book.js]: Yeah, that is true, do you want books without Author to be published?

File: .eslintrc (hunk 6 above)
Comment (emp-daisy) [reply .eslintrc]: Then why did you format it? Revert to what it was, since no change intended

File: server/controllers/bookController.js (hunk 8 above)
Comment (PelumiAlesh) [reply server/controllers/bookController.js:21]: I do not understand this @emp-daisy

File: server/db/migrations/20190605110045-create-book.js (hunk 7 above)
Comment (emp-daisy) [reply server/db/migrations/20190605110045-create-book.js]: There's a possibility that might happen but then a relationship has to exist between the books and author

Comment (PelumiAlesh) [reply server/db/migrations/20190605110045-create-book.js]: True that, do you want me to change it regardlessly?

**Review by PelumiAlesh — COMMENTED** [review]
okay

File: server/controllers/bookController.js (hunk 8 above)
Comment (emp-daisy) [reply server/controllers/bookController.js:21]: I mean what happens when an ID that doesn't exist is passed?

Comment (PelumiAlesh) [reply server/controllers/bookController.js:21]: Done! I made it possible to post book without Author

Comment (emp-daisy) [reply server/controllers/bookController.js:21]: That's not the only issue.
Imagine a scenario where a user provides ID "20" for the authorID , and there is no author with such ID. What happens in this case?

https://sequelize.readthedocs.io/en/latest/docs/associations/
http://docs.sequelizejs.com/manual/associations.html

**Review by OvieMudi — DISMISSED** [review]

File: server/docs/hellobooks_api_doc.yaml [hunk 9]
```diff
@@ -202,6 +204,103 @@ paths:
...
+            application/json:
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+        500:
+          description: Server error
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+=======
```
Comment (emp-daisy) [inline server/docs/hellobooks_api_doc.yaml]: Fix this issue from resolving conflict

File: server/docs/hellobooks_api_doc.yaml [hunk 10]
```diff
@@ -202,6 +204,103 @@ paths:
...
+            application/json:
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+        500:
+          description: Server error
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+<<<<<<< HEAD
```
Comment (emp-daisy) [inline server/docs/hellobooks_api_doc.yaml]: Here too

File: server/docs/hellobooks_api_doc.yaml [hunk 11]
```diff
@@ -202,6 +204,103 @@ paths:
...
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+        500:
+          description: Server error
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/errorResponse"
+=======
+>>>>>>> 6d352527e0a05868791b7235c9daccd5e02c7b08
```
Comment (emp-daisy) [inline server/docs/hellobooks_api_doc.yaml]: Remove this

**Review by emp-daisy — DISMISSED** [review]

**Review by emp-daisy — APPROVED** [review]

**Review by Hadeneekeh — APPROVED** [review]

**Review by OvieMudi — APPROVED** [review]
